README
GraphQL

A reference implementation of the GraphQL spec in Perl.

INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc GraphQL

You can also look for information at:

    RT, CPAN's request tracker (report bugs here)
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=GraphQL

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/GraphQL

    CPAN Ratings
        http://cpanratings.perl.org/d/GraphQL

    Search CPAN
        http://search.cpan.org/dist/GraphQL/


LICENSE AND COPYRIGHT

Copyright (C) 2017 Ed J

This program is free software; you can redistribute it and/or modify it
under the terms of the the Artistic License (2.0). You may obtain a
copy of the full license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any use, modification, and distribution of the Standard or Modified
Versions is governed by this Artistic License. By using, modifying or
distributing the Package, you accept this license. Do not use, modify,
or distribute the Package, if you do not accept this license.

If your Modified Version has been derived from a Modified Version made
by someone other than you, you are nevertheless required to ensure that
your Modified Version complies with the requirements of this license.

This license does not grant you the right to use any trademark, service
mark, tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge
patent license to make, have made, use, offer to sell, sell, import and
otherwise transfer the Package with respect to any patent claims
licensable by the Copyright Holder that are necessarily infringed by the
Package. If you institute patent litigation (including a cross-claim or
counterclaim) against any party alleging that the Package constitutes
direct or contributory patent infringement, then this Artistic License
to you shall terminate on the date that such litigation is filed.

Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR
CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR
CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.



Changes
0.54 2022-03-24
- replace use of Type::Tiny assertions for performance reasons

0.53 2022-01-15
- String and ID now always serialise to JSON strings - thanks @cliveholloway

0.52 2021-09-03
- not always load DateTime type - thanks @mangano-ito

0.51 2021-07-04
- from_doc accepts kind2class to override SDL keyword - thanks @mangano-ito

0.50 2021-04-08
- Scalar.parse_value was wrong for Boolean - thanks @ccakes for report

0.49 2021-02-15
- type-checking only on direct MaybeTypeCheck use (#35) - thanks @ccakes

0.48 2021-02-05
- require Import::Into 1.002003 as we depend on module-loading

0.47 2021-02-02
- type-checking now only when Devel::StrictMode STRICT (#31) - thanks @ccakes
- can give default values to enums in queries (#33) - thanks @ccakes
- eliminate tail-call in type expansion - thanks @ccakes

0.46 2021-01-27
- PromiseCode example to interoperate with Promises - thanks @hitode909

0.45 2021-01-27
- Promise only needs then method - resolve/reject go via PromiseCode - thanks @hitode909 for prompt

0.44 2020-11-09
- Allow serialising already-blessed bools (#30) - thanks @ccakes

0.43 2020-11-07
- mutations now guaranteed to be executed in order - thanks @ccakes for prompt

0.42 2020-11-07
- List needed a perl_to_graphql method - thanks @schoeer for report
- so did InputObject

0.41 2020-08-05
- Scalar types now correctly handle null values

0.40 2020-05-09
- GraphQL::AsyncIterator->map_then now mutates object, returns self

0.39 2020-05-06
- simplify Int and Float implementations
- improve Boolean error-checking
- Convert::Test has an "echo" mutation and optional "timedEcho" subscription
- Subscriptions implemented

0.38 2019-12-08
- error objects now stringify - thanks @jjn1056 for idea
- more tests for directives - thanks @jjn1056
- Support field directives - thanks @yigarashi-9

0.37 2019-11-04
 - fix literal input-object coercion problem - thanks @ccakes
 - add "extensions" field to GraphQL::Error - thanks @cockscomb

0.36 2019-05-01
 - doc improvements

0.35 2019-04-29
 - remove author-mode stuff from affecting make in distro

0.34 2019-04-28
 - List now has "name" that is its "of"'s name

0.33	Thu 21 Feb 22:58:02 GMT 2019
 - update Plugin::Convert docs - thanks @aksonov
 - zap "meaningful comment" as description
 - implement GraphQL::Plugin::Type API, switch DateTime to that

0.32	Tue  5 Feb 02:37:20 GMT 2019
 - metadata and test updates
 - unescape string literals - thanks @chazmcgarvey
 - add block strings - thanks @chazmcgarvey
 - update SDL parsing to current graphql-js, including empty defs of most types
 - make to_doc produce new-style descriptions while from_doc still understands old, so can convert

0.31	Sun  1 Jul 23:10:59 BST 2018
 - Fix non-nullable Enum being misdetected - thanks @mechairoi

0.30	Wed 28 Feb 04:29:50 GMT 2018
 - make variable-type errors JSONable for returning - thanks @nohuhu

0.29	Thu Feb 15 03:40:12 GMT 2018
 - Pegex version that gets line_column right

0.28	Sat Feb 10 18:26:29 GMT 2018
 - fix Enum type-check bug #10

0.27	Fri Feb  9 18:13:35 GMT 2018
 - fix input type-check bug #9

0.26	Sat Dec 30 08:45:37 GMT 2017
 - the "then" of ->all will get each result as array-ref

0.25	Fri Dec 29 12:47:13 GMT 2017
 - more-correct handle of lists and non-null
 - handles Promises if give helper code-refs (type PromiseCode)

0.24	Tue 12 Dec 03:57:18 GMT 2017
 - handle array of enum

0.23	Sat  9 Dec 17:48:42 GMT 2017
 - fix 5.14 bug with "?" ambiguity

0.22	Sat  9 Dec 06:44:51 GMT 2017
 - basic Dockerfile makes 5.26.1 image with GraphQL
 - multi-line descriptions parsed and round-tripped right
 - doc updates

0.21	Thu Nov 16 05:22:53 GMT 2017
 - add DateTime scalar type
 - allow class overrides in Schema.from_ast
 - add and use _complete_value to the Type classes

0.20	Wed Nov  1 03:59:54 GMT 2017
 - add GraphQL::Plugin::Convert API

0.19	Sun Oct 29 16:11:31 GMT 2017
 - add DateTime as test dep, thanks @andk!

0.18	Fri Oct 27 05:00:46 BST 2017
 - make AST not have "node"

0.17	Thu Oct 26 19:58:00 BST 2017
 - fill out Schema.to_doc include default if eg type Query

0.16	Sun Oct 22 07:09:44 BST 2017
 - execute take AST as well as text
 - error reporting improvements
 - implement @skip and @include
 - implement GraphQL::Schema->from_doc et al
 - rework of graphql.pgx

0.15	Mon Oct 16 00:13:31 BST 2017
 - introspection improvements
 - switch parse and execute from methods to functions

0.14	Fri Oct 13 06:31:33 BST 2017
 - catch and report query-parse errors
 - parse and other errors now report query-doc location
 - eliminated various data mutations for purer functions

0.13	Thu Oct 12 16:18:20 BST 2017
 - bulletproof JSON booleans. figured with help from @haarg and @shadowcat-mst

0.12	Tue Oct 10 05:13:29 BST 2017
 - allow $context to be anything (such as request headers)

0.11	Mon Oct  9 17:17:17 BST 2017
 - coerce numbers in Int and Float to dodge stringifying of numbers for JSON::XS

0.10	Thu Oct  5 05:30:56 BST 2017
 - tighten the JSON module requirements to zap stringifying-numbers test fails

0.09	Mon Oct  2 06:39:07 BST 2017
 - implement conversion between Perl and GraphQL/JSON values
 - add more usable debugging code
 - error handling
 - implement Object field resolving
 - implement fragments
 - implement Enum, Directive
 - more type checking
 - implement introspection
 - list GitHub as bugtracker
 - documentation

0.08	Sat Sep 23 06:27:20 BST 2017
 - zap GraphQL::Argument
 - get all thunking right using MooX::Thunking
 - is_deprecated auto handling
 - Execution does top-level field resolving, type coercing
 - port execution-resolve.t execution-variables.t from graphql-js
 - Schema API mainly complete
 - many types for validation
 - make Parser return usable AST
 - improvements to Pegex grammar from ingy
 - start of GraphQL::Error
 - JSON::MaybeXS so no stringify numbers

0.07	Tue Sep  5 19:50:33 BST 2017
 - make GraphQL::Directive, GraphQL::Argument
 - make GraphQL::Type::Library with Type::Tiny constraints

Makefile.PL
use 5.014;
use strict;
use warnings;
use ExtUtils::MakeMaker;

WriteMakefile(
  NAME             => 'GraphQL',
  AUTHOR           => q{Ed J <etj@cpan.org>},
  VERSION_FROM     => 'lib/GraphQL.pm',
  ABSTRACT_FROM    => 'lib/GraphQL.pm',
  LICENSE          => 'artistic_2',
  MIN_PERL_VERSION => '5.014',
  CONFIGURE_REQUIRES => {
    'ExtUtils::MakeMaker' => '6.64', # TEST_REQUIRES
  },
  TEST_REQUIRES => {
    'Test::More' => '0.88', # done_testing
    'Test::Exception' => '0.42',
    'Test::Deep' => '1.127',
  },
  PREREQ_PM => {
    'Moo' => '0',
    'Attribute::Handlers' => '0',
    'Import::Into' => '1.002003', # loads modules
    'Type::Tiny' => '0',
    'Module::Runtime' => '0',
    'Function::Parameters' => '2.001001',
    'Return::Type' => '0',
    'Pegex' => '0.64',
    'MooX::Thunking' => '0.07', # takes care of Thunk in isa
    'JSON::MaybeXS' => '1.003009', # is_bool, also . @INC
    'JSON::PP' => '2.92', # number detection
    'DateTime' => 0,
    'DateTime::Format::ISO8601' => 0,
    'Devel::StrictMode' => 0,
    'curry' => 0,
  },
  dist  => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
  clean => { FILES => 'GraphQL-*' },
  META_MERGE => {
    "meta-spec" => { version => 2 },
    dynamic_config => 0,
    resources => {
      x_IRC => 'irc://irc.perl.org/#graphql-perl',
      repository => {
        type => 'git',
        url => 'git@github.com:graphql-perl/graphql-perl.git',
        web => 'https://github.com/graphql-perl/graphql-perl',
      },
      bugtracker  => {
        web => 'https://github.com/graphql-perl/graphql-perl/issues',
      },
      license => [ 'http://dev.perl.org/licenses/' ],
    },
    prereqs => {
      develop => {
        requires => {
          'Test::Pod::Coverage' => '1.08',
          'Test::Pod' => '1.22',
          'Pod::Markdown' => 0,
        },
      },
      runtime => {
        suggests => {
          'Cpanel::JSON::XS' => '3.0237',
          'JSON::XS' => 0,
        },
      },
    },
  },
);

sub MY::postamble {
  return '' if !-e '.git';
  my $container = 'graphqlperl/graphql';
  my $prereq = "${container}-prereq";
  <<EOF;
\$(VERSION_FROM) : graphql.pgx
\t\$(PERL) -Ilib -MGraphQL::Language::Grammar=compile

pure_all :: README.md

README.md : \$(VERSION_FROM)
\tpod2markdown \$< >\$\@

containerprereq :
\tdocker build -f Dockerfile.prereq -t $prereq .

containerprereqpush :
\tdocker push $prereq

container :
\tdocker build -t $container:\$(VERSION) .

containerpush :
\tdocker push $container:\$(VERSION)
EOF
}
